<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Oliver Y. Hasegawa">

<title>Bayesian Statistical Modeling Examples</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="bayesian_statistical_modeling_examples_files/libs/clipboard/clipboard.min.js"></script>
<script src="bayesian_statistical_modeling_examples_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="bayesian_statistical_modeling_examples_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="bayesian_statistical_modeling_examples_files/libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="bayesian_statistical_modeling_examples_files/libs/quarto-html/popper.min.js"></script>
<script src="bayesian_statistical_modeling_examples_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="bayesian_statistical_modeling_examples_files/libs/quarto-html/anchor.min.js"></script>
<link href="bayesian_statistical_modeling_examples_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="bayesian_statistical_modeling_examples_files/libs/quarto-html/quarto-syntax-highlighting-8d579c7e4cb5e38eaddfdf7e7f03f655.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="bayesian_statistical_modeling_examples_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="bayesian_statistical_modeling_examples_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="bayesian_statistical_modeling_examples_files/libs/bootstrap/bootstrap-c0728b2b96de12ceda148716bf85ba50.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN" && texText && texText.data) {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#instruction" id="toc-instruction" class="nav-link active" data-scroll-target="#instruction">Instruction</a></li>
  <li><a href="#multivariate-normal-distribution-model" id="toc-multivariate-normal-distribution-model" class="nav-link" data-scroll-target="#multivariate-normal-distribution-model">1. Multivariate normal distribution model</a>
  <ul class="collapse">
  <li><a href="#mathematical-equation" id="toc-mathematical-equation" class="nav-link" data-scroll-target="#mathematical-equation">1.1 Mathematical equation</a></li>
  <li><a href="#stan-code" id="toc-stan-code" class="nav-link" data-scroll-target="#stan-code">1.2 Stan code</a></li>
  </ul></li>
  <li><a href="#bayesian-linear-regression-model" id="toc-bayesian-linear-regression-model" class="nav-link" data-scroll-target="#bayesian-linear-regression-model">2. Bayesian linear regression model</a>
  <ul class="collapse">
  <li><a href="#mathematical-equation-1" id="toc-mathematical-equation-1" class="nav-link" data-scroll-target="#mathematical-equation-1">2.1 Mathematical equation</a></li>
  <li><a href="#stan-code-1" id="toc-stan-code-1" class="nav-link" data-scroll-target="#stan-code-1">2.2 Stan code</a></li>
  </ul></li>
  <li><a href="#bayesian-hierarchical-random-intercept-model" id="toc-bayesian-hierarchical-random-intercept-model" class="nav-link" data-scroll-target="#bayesian-hierarchical-random-intercept-model">3. Bayesian Hierarchical Random Intercept model</a>
  <ul class="collapse">
  <li><a href="#mathematical-equation-2" id="toc-mathematical-equation-2" class="nav-link" data-scroll-target="#mathematical-equation-2">3.1 Mathematical equation</a></li>
  <li><a href="#stan-code-2" id="toc-stan-code-2" class="nav-link" data-scroll-target="#stan-code-2">3.2 Stan code</a></li>
  </ul></li>
  <li><a href="#bayesian-hierarchical-random-coefficiant-model-with-correlation-matrix" id="toc-bayesian-hierarchical-random-coefficiant-model-with-correlation-matrix" class="nav-link" data-scroll-target="#bayesian-hierarchical-random-coefficiant-model-with-correlation-matrix">4. Bayesian Hierarchical Random Coefficiant Model with Correlation Matrix</a>
  <ul class="collapse">
  <li><a href="#mathematical-equation-3" id="toc-mathematical-equation-3" class="nav-link" data-scroll-target="#mathematical-equation-3">4.1 Mathematical equation</a></li>
  <li><a href="#stan-code-3" id="toc-stan-code-3" class="nav-link" data-scroll-target="#stan-code-3">4.2 Stan code</a></li>
  </ul></li>
  <li><a href="#gaussian-process-model-with-phylogenetic-distance-rbf-kernel" id="toc-gaussian-process-model-with-phylogenetic-distance-rbf-kernel" class="nav-link" data-scroll-target="#gaussian-process-model-with-phylogenetic-distance-rbf-kernel">5. Gaussian Process Model with Phylogenetic distance RBF kernel</a>
  <ul class="collapse">
  <li><a href="#mathematical-equation-4" id="toc-mathematical-equation-4" class="nav-link" data-scroll-target="#mathematical-equation-4">5.1 Mathematical equation</a></li>
  <li><a href="#stan-code-4" id="toc-stan-code-4" class="nav-link" data-scroll-target="#stan-code-4">5.2 Stan code</a></li>
  </ul></li>
  <li><a href="#additional-hierarchical-bayesian-linear-model-with-random-effects-covariance-structure" id="toc-additional-hierarchical-bayesian-linear-model-with-random-effects-covariance-structure" class="nav-link" data-scroll-target="#additional-hierarchical-bayesian-linear-model-with-random-effects-covariance-structure">Additional: Hierarchical Bayesian Linear Model with Random-Effects Covariance Structure</a>
  <ul class="collapse">
  <li><a href="#mathematical-equation-5" id="toc-mathematical-equation-5" class="nav-link" data-scroll-target="#mathematical-equation-5">Mathematical equation</a></li>
  <li><a href="#brms-code" id="toc-brms-code" class="nav-link" data-scroll-target="#brms-code">brms code</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Bayesian Statistical Modeling Examples</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Oliver Y. Hasegawa </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="instruction" class="level1">
<h1>Instruction</h1>
<p>Thank you very much for tuning in.<br>
Bellow are examples of Bayesian statistical modeling with mathematical equation and its Stan coding (There’s also additional model using brms package in R).<br>
Here we have variables called “<strong>tightness</strong>” and “<strong>X1</strong>” in our models.</p>
</section>
<section id="multivariate-normal-distribution-model" class="level1">
<h1>1. Multivariate normal distribution model</h1>
<section id="mathematical-equation" class="level2">
<h2 class="anchored" data-anchor-id="mathematical-equation">1.1 Mathematical equation</h2>
<p><span class="math display">\[
\begin{pmatrix}
\text{tightness}_i \\
X1_i
\end{pmatrix}
\sim \mathcal{N}\!\left(
\begin{pmatrix}
\mu_{\text{tight}} \\
\mu_{X1}
\end{pmatrix},
\begin{pmatrix}
\sigma^2_{\text{tight}} &amp; \rho \,\sigma_{\text{tight}}\sigma_{X1} \\
\rho \,\sigma_{\text{tight}}\sigma_{X1} &amp; \sigma^2_{X1}
\end{pmatrix}
\right)
\]</span></p>
<p>i represents the index of countries or states.<br>
<span class="math inline">\(tightness\)</span> and <span class="math inline">\(X_1\)</span> represent the variables which follow a bivariate normal distribution.<br>
<span class="math inline">\(\mu_{\text{tight}}\)</span> and <span class="math inline">\(\mu_{X1}\)</span> respectively represent the means of <span class="math inline">\(tightness_i\)</span> and <span class="math inline">\(X_{1i}\)</span>.<br>
<span class="math inline">\(\sigma^2_{\text{tight}}\)</span> and <span class="math inline">\(\sigma^2_{X1}\)</span> are the variances of <span class="math inline">\(tightness_i\)</span> and <span class="math inline">\(X_{1i}\)</span>.<br>
<span class="math inline">\(\rho\)</span> is the correlation coefficient between <span class="math inline">\(tightness_i\)</span> and <span class="math inline">\(X_{1i}\)</span>.<br>
We assumed independence across <span class="math inline">\(i\)</span> observations and estimated the parameters<br>
<span class="math inline">\(\mu_{\text{tight}}\)</span>, <span class="math inline">\(\mu_{X1}\)</span>, <span class="math inline">\(\sigma^2_{\text{tight}}\)</span>, <span class="math inline">\(\sigma^2_{X1}\)</span>, and <span class="math inline">\(\rho\)</span>.</p>
</section>
<section id="stan-code" class="level2">
<h2 class="anchored" data-anchor-id="stan-code">1.2 Stan code</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="sourceCode stan code-with-copy"><code class="sourceCode stan"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="kw">data</span> {</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>&gt; N;</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="dt">array</span>[N] <span class="dt">vector</span>[<span class="dv">2</span>] y;</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[<span class="dv">2</span>] mu_prior_mean;</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt;[<span class="dv">2</span>] mu_prior_sd;</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt;[<span class="dv">2</span>] sigma_prior;</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>&gt; eta;</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="kw">parameters</span> {</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[<span class="dv">2</span>] mu;</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt;[<span class="dv">2</span>] sigma;</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>  <span class="dt">cholesky_factor_corr</span>[<span class="dv">2</span>] rho;</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="kw">transformed parameters</span> {</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>  <span class="dt">matrix</span>[<span class="dv">2</span>,<span class="dv">2</span>] L_Sigma = diag_pre_multiply(sigma, rho);</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="kw">model</span> {</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>  mu ~ normal(mu_prior_mean, mu_prior_sd);</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>  sigma ~ exponential(sigma_prior);</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>  rho ~ lkj_corr_cholesky(eta);</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>    y ~ multi_normal_cholesky(mu, L_Sigma);</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a><span class="kw">generated quantities</span> {</span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>  <span class="dt">array</span>[N] <span class="dt">vector</span>[<span class="dv">2</span>] y_rep;</span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span>:N) {</span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a>    y_rep[n] = multi_normal_cholesky_rng(mu, L_Sigma);</span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
</section>
<section id="bayesian-linear-regression-model" class="level1">
<h1>2. Bayesian linear regression model</h1>
<section id="mathematical-equation-1" class="level2">
<h2 class="anchored" data-anchor-id="mathematical-equation-1">2.1 Mathematical equation</h2>
<p><span class="math display">\[
\begin{aligned}
\text{tightness}_i &amp;\sim \text{Normal}(\mu_i, \sigma) \\
\mu_i &amp;= \beta_0 + \beta_1 X1_i \\[6pt]
\beta_0 &amp;\sim \text{Normal}(\mu_{\beta_0}, \tau_{\beta_0}) \\
\beta_1 &amp;\sim \text{Normal}(\mu_{\beta_1}, \tau_{\beta_1}) \\
\sigma &amp;\sim \text{Exponential}(\lambda)
\end{aligned}
\]</span></p>
<p><span class="math inline">\(tightness_i\)</span> is predicted by <span class="math inline">\(X_{1i}\)</span>, and <span class="math inline">\(\mu_i\)</span> is the expected value of <span class="math inline">\(tightness_i\)</span>.<br>
While <span class="math inline">\(\beta_0\)</span> represents the intercept, <span class="math inline">\(\beta_1\)</span> represents the slope coefficient.<br>
<span class="math inline">\(\sigma\)</span> is the residual standard deviation.</p>
<p>It is assumed that <span class="math inline">\(\beta_0\)</span> and <span class="math inline">\(\beta_1\)</span> are sampled from normal distributions whose means and standard deviations are <span class="math inline">\(\mu_\beta\)</span> and <span class="math inline">\(\tau_\beta\)</span>, respectively.<br>
The values of <span class="math inline">\(X_1\)</span>, <span class="math inline">\(\beta_0\)</span>, <span class="math inline">\(\beta_1\)</span>, <span class="math inline">\(\mu_\beta\)</span>, <span class="math inline">\(\tau_\beta\)</span>, and <span class="math inline">\(\sigma\)</span> were estimated as parameters.</p>
</section>
<section id="stan-code-1" class="level2">
<h2 class="anchored" data-anchor-id="stan-code-1">2.2 Stan code</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="sourceCode stan code-with-copy"><code class="sourceCode stan"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="kw">data</span> {</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>&gt; N;</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[N] X1;</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[N] tightness;</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>, <span class="kw">upper</span>=<span class="dv">1</span>&gt; prior_only;</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> mu_alpha;</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; tau_alpha;</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> mu_beta;</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; tau_beta;</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; lambda;</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="kw">parameters</span> {</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> alpha;</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> beta;</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; sigma;</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="kw">model</span> {</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a> <span class="dt">vector</span>[N] mu;</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a> mu = alpha + beta*X1;</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>  alpha ~ normal(mu_alpha, tau_alpha);</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>  beta ~ normal(mu_beta, tau_beta);</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>  sigma ~ exponential(lambda);</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (prior_only == <span class="dv">0</span>) {</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>  tightness ~ normal(mu, sigma);</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a><span class="kw">generated quantities</span> {</span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[N] y_ppc;</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span>:N) {</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>    y_ppc[n] = normal_rng(alpha + beta * X1[n], sigma);</span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
</section>
<section id="bayesian-hierarchical-random-intercept-model" class="level1">
<h1>3. Bayesian Hierarchical Random Intercept model</h1>
<section id="mathematical-equation-2" class="level2">
<h2 class="anchored" data-anchor-id="mathematical-equation-2">3.1 Mathematical equation</h2>
<p><span class="math display">\[
\begin{aligned}
\text{tightness}_i &amp;\sim \text{Normal}(\mu_i, \sigma) \\
\mu_i &amp;= \alpha_{\text{Region}[i]} + \beta_1 X1_i \\[10pt]
\alpha_{\text{Region}} &amp;\sim \text{Normal}(\mu_\alpha, \tau_\alpha) \\
\beta_1 &amp;\sim \text{Normal}(\mu_{\beta_1}, \tau_{\beta_1}) \\
\mu_\alpha &amp;\sim \text{Normal}(\mu_0, \tau_0) \\
\tau_\alpha &amp;\sim \text{Exponential}(\lambda_\alpha) \\
\sigma &amp;\sim \text{Exponential}(\lambda_\sigma)
\end{aligned}
\]</span></p>
<p><span class="math inline">\(tightness_i\)</span> is predicted by <span class="math inline">\(X_{1i}\)</span>, and <span class="math inline">\(\mu_i\)</span> is the expected value of <span class="math inline">\(tightness_i\)</span>.<br>
<span class="math inline">\(\sigma\)</span> is the residual standard deviation.<br>
<span class="math inline">\(\alpha_{\text{Region}[i]}\)</span> is the intercept specific to each <span class="math inline">\(\text{Region}[i]\)</span>, and <span class="math inline">\(\beta_1\)</span> is the fixed-effect slope coefficient.<br>
<span class="math inline">\(\mu_a\)</span> and <span class="math inline">\(\tau_a\)</span> are the hyperparameters of <span class="math inline">\(\alpha_{\text{Region}}\)</span>.</p>
<p>The values of <span class="math inline">\(X_1\)</span>, <span class="math inline">\(\alpha_{\text{Region}}\)</span>, <span class="math inline">\(\beta_1\)</span>, <span class="math inline">\(\mu_a\)</span>, <span class="math inline">\(\tau_a\)</span>, <span class="math inline">\(\mu_\beta\)</span>, <span class="math inline">\(\tau_\beta\)</span>, <span class="math inline">\(\mu_0\)</span>, <span class="math inline">\(\tau_0\)</span>, and <span class="math inline">\(\sigma\)</span> were estimated as parameters.<br>
<span class="math inline">\(\mu_a\)</span>, <span class="math inline">\(\mu_\beta\)</span>, and <span class="math inline">\(\mu_0\)</span> were assigned normal distributions, while <span class="math inline">\(\tau_a\)</span> and <span class="math inline">\(\sigma\)</span> were assigned exponential distributions.</p>
</section>
<section id="stan-code-2" class="level2">
<h2 class="anchored" data-anchor-id="stan-code-2">3.2 Stan code</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="sourceCode stan code-with-copy"><code class="sourceCode stan"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="kw">data</span> {</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>&gt; N;</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>&gt; R;</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="dt">array</span>[N] <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>, <span class="kw">upper</span>=R&gt; region;</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[N] X1;</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[N] tightness;</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>, <span class="kw">upper</span>=<span class="dv">1</span>&gt; prior_only;</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> mu_alpha_bar;</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> tau_alpha_bar;</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> mu_beta;</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> tau_beta;</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; lambda_alpha;</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; lambda_sigma;</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="kw">parameters</span> {</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> alpha_bar;</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; alpha_sigma;</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[R] alpha_raw;</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> beta;</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; sigma;</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a><span class="kw">transformed parameters</span> {</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[R] alpha;</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>  alpha = alpha_bar + alpha_sigma * alpha_raw;</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a><span class="kw">model</span> {</span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>  beta ~ normal(mu_beta, tau_beta);</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a>  alpha_bar ~ normal(mu_alpha_bar, tau_alpha_bar);</span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a>  alpha_sigma ~ exponential(lambda_alpha);</span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a>  alpha_raw ~ normal(<span class="dv">0</span>, <span class="dv">1</span>);</span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a>  sigma ~ exponential(lambda_sigma);</span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (prior_only == <span class="dv">0</span>) {</span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>    tightness ~ normal(alpha[region] + beta * X1, sigma);</span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a>   }</span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a><span class="kw">generated quantities</span> {</span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[N] y_ppc;</span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span>:N) {</span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true" tabindex="-1"></a>    y_ppc[i] = normal_rng(alpha[region[i]] + beta * X1[i], sigma);</span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb3-47"><a href="#cb3-47" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
</section>
<section id="bayesian-hierarchical-random-coefficiant-model-with-correlation-matrix" class="level1">
<h1>4. Bayesian Hierarchical Random Coefficiant Model with Correlation Matrix</h1>
<section id="mathematical-equation-3" class="level2">
<h2 class="anchored" data-anchor-id="mathematical-equation-3">4.1 Mathematical equation</h2>
<p><span class="math display">\[
\begin{aligned}
\text{tightness}_i &amp;\sim \text{Normal}(\mu_i, \sigma) \\
\mu_i &amp;= \alpha_{\text{Region}[i]} + \beta_{\text{Region}[i]} X1_i \\[10pt]
\begin{bmatrix}
\alpha_{\text{Region}} \\
\beta_{\text{Region}}
\end{bmatrix}
&amp;\sim \text{MVNormal}
\left(
\begin{bmatrix}
\alpha_0 \\
\beta_0
\end{bmatrix},
\mathbf{S}
\right) \\[10pt]
\mathbf{S} &amp;=
\begin{pmatrix}
\sigma_\alpha &amp; 0 \\
0 &amp; \sigma_\beta
\end{pmatrix}
\mathbf{R}
\begin{pmatrix}
\sigma_\alpha &amp; 0 \\
0 &amp; \sigma_\beta
\end{pmatrix} \\[10pt]
\alpha_0 &amp;\sim \text{Normal}(\mu_{\alpha_0}, \tau_{\alpha_0}) \\
\beta_0 &amp;\sim \text{Normal}(\mu_{\beta_0}, \tau_{\beta_0}) \\
\sigma_\alpha &amp;\sim \text{Exponential}(\lambda_\alpha) \\
\sigma_\beta &amp;\sim \text{Exponential}(\lambda_\beta) \\
\mathbf{R} &amp;\sim \text{LKJcorr}(\eta) \\
\sigma &amp;\sim \text{Exponential}(\lambda_\sigma)
\end{aligned}
\]</span></p>
<p><span class="math inline">\(\alpha_{\text{Region}[i]}\)</span> is the region-specific intercept, and <span class="math inline">\(\beta_{\text{Region}[i]}\)</span> is the region-specific slope for predictor <span class="math inline">\(X_{1i}\)</span>.<br>
<span class="math inline">\(\mu_i\)</span> is the expected value of <span class="math inline">\(tightness_i\)</span>, and <span class="math inline">\(\sigma\)</span> is the residual standard deviation.</p>
<p>We modeled the random effects <span class="math inline">\(\alpha_{\text{Region}}\)</span> and <span class="math inline">\(\beta_{\text{Region}}\)</span> jointly as draws from a bivariate normal distribution.<br>
The covariance matrix <span class="math inline">\(\mathbf{S}\)</span> was decomposed into standard deviations and a correlation matrix <span class="math inline">\(\mathbf{R}\)</span>.<br>
<span class="math inline">\(\mathbf{R} \sim \mathrm{LKJcorr}(\eta)\)</span> represents the correlation matrix between the region-specific intercepts and slopes,<br>
while <span class="math inline">\(\sigma_\alpha\)</span> and <span class="math inline">\(\sigma_\beta\)</span> are their respective standard deviations.</p>
<p>The values of <span class="math inline">\(X_1\)</span>, <span class="math inline">\(\alpha_{\text{Region}}\)</span>, <span class="math inline">\(\beta_{\text{Region}}\)</span>, <span class="math inline">\(\alpha_0\)</span>, <span class="math inline">\(\beta_0\)</span>, <span class="math inline">\(\mu_a\)</span>, <span class="math inline">\(\tau_a\)</span>, <span class="math inline">\(\mu_\beta\)</span>, <span class="math inline">\(\tau_\beta\)</span>, <span class="math inline">\(\mu_0\)</span>, <span class="math inline">\(\tau_0\)</span>, and <span class="math inline">\(\sigma\)</span> were estimated as parameters.<br>
<span class="math inline">\(\alpha_0\)</span> and <span class="math inline">\(\beta_0\)</span> were assigned normal distributions.<br>
<span class="math inline">\(\sigma_\alpha\)</span>, <span class="math inline">\(\sigma_\beta\)</span>, and <span class="math inline">\(\sigma\)</span> were assigned exponential distributions.</p>
</section>
<section id="stan-code-3" class="level2">
<h2 class="anchored" data-anchor-id="stan-code-3">4.2 Stan code</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb4"><pre class="sourceCode stan code-with-copy"><code class="sourceCode stan"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="kw">data</span> {</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>&gt; N;</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>&gt; R;</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="dt">array</span>[N] <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>, <span class="kw">upper</span>=R&gt; region;</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[N] X1;</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[N] tightness;</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>,<span class="kw">upper</span>=<span class="dv">1</span>&gt; prior_only;</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> mu_alpha;</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> tau_alpha;</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> mu_beta;</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> tau_beta;</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; lambda_alpha;</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; lambda_beta;</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; lambda_sigma;</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="kw">parameters</span> {</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> a;</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> b;</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; sigma;</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt;[<span class="dv">2</span>] sigma_region;</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>  <span class="dt">cholesky_factor_corr</span>[<span class="dv">2</span>] L_Rho;</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>  <span class="dt">matrix</span>[<span class="dv">2</span>, R] z;</span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a><span class="kw">transformed parameters</span> {</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>  <span class="dt">matrix</span>[<span class="dv">2</span>, R] dev = diag_pre_multiply(sigma_region, L_Rho) * z;</span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[R] a_region = rep_vector(a, R) + dev[<span class="dv">1</span>]';</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[R] b_region = rep_vector(b, R) + dev[<span class="dv">2</span>]';</span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a><span class="kw">model</span> {</span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[N] mu;</span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a>  a ~ normal(mu_alpha, tau_alpha);</span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a>  b ~ normal(mu_beta,  tau_beta);</span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a>  sigma ~ exponential(lambda_sigma);</span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a>  sigma_region[<span class="dv">1</span>] ~ exponential(lambda_alpha);</span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a>  sigma_region[<span class="dv">2</span>] ~ exponential(lambda_beta);</span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a>  L_Rho ~ lkj_corr_cholesky(<span class="dv">4</span>);</span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a>  to_vector(z) ~ normal(<span class="dv">0</span>, <span class="dv">1</span>);</span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span>:N) {</span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a>    mu[n] = a_region[region[n]] + b_region[region[n]] * X1[n];</span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (prior_only == <span class="dv">0</span>) {</span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a>  tightness ~ normal(mu, sigma);</span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb4-52"><a href="#cb4-52" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-53"><a href="#cb4-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-54"><a href="#cb4-54" aria-hidden="true" tabindex="-1"></a><span class="kw">generated quantities</span> {</span>
<span id="cb4-55"><a href="#cb4-55" aria-hidden="true" tabindex="-1"></a>  <span class="dt">corr_matrix</span>[<span class="dv">2</span>] Rho_out = multiply_lower_tri_self_transpose(L_Rho);</span>
<span id="cb4-56"><a href="#cb4-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-57"><a href="#cb4-57" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[N] y_ppc;</span>
<span id="cb4-58"><a href="#cb4-58" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span>:N)</span>
<span id="cb4-59"><a href="#cb4-59" aria-hidden="true" tabindex="-1"></a>    y_ppc[n] = normal_rng(a_region[region[n]] + b_region[region[n]] * X1[n], sigma);</span>
<span id="cb4-60"><a href="#cb4-60" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
</section>
<section id="gaussian-process-model-with-phylogenetic-distance-rbf-kernel" class="level1">
<h1>5. Gaussian Process Model with Phylogenetic distance RBF kernel</h1>
<section id="mathematical-equation-4" class="level2">
<h2 class="anchored" data-anchor-id="mathematical-equation-4">5.1 Mathematical equation</h2>
<p><span class="math display">\[
\begin{aligned}
\text{tightness}_i &amp;\sim \text{Normal}(\mu_i, \sigma) \\
\mu_i &amp;= \alpha + \gamma_{\text{Phylogeny}[i]} + \beta X1_i \\[10pt]
\boldsymbol{\gamma} &amp;\sim \text{MVNormal}(\mathbf{0}, \mathbf{K}) \\
K_{ij} &amp;= \eta^2 \exp\!\left( -\rho^2 D_{ij}^2 \right) + \delta_{ij}\epsilon \\[10pt]
\alpha &amp;\sim \text{Normal}(\mu_\alpha, \tau_\alpha) \\
\beta &amp;\sim \text{Normal}(\mu_\beta, \tau_\beta) \\
\eta &amp;\sim \text{Exponential}(\lambda_\eta) \\
\rho &amp;\sim \text{Exponential}(\lambda_\rho) \\
\sigma &amp;\sim \text{Exponential}(\lambda_\sigma)
\end{aligned}
\]</span></p>
<p><span class="math inline">\(tightness_i\)</span> is the observed outcome for the national unit, and <span class="math inline">\(X_{1i}\)</span> is the predictor variable.<br>
<span class="math inline">\(\alpha\)</span> is the intercept, and <span class="math inline">\(\beta\)</span> is the fixed-effect slope.</p>
<p>Importantly, <span class="math inline">\(\gamma_{\text{Phylogeny}[i]}\)</span> is a structured random effect drawn from a Gaussian Process, and <span class="math inline">\(\sigma\)</span> is the residual standard deviation.<br>
The vector of structured effects <span class="math inline">\(\boldsymbol{\gamma}\)</span> is modeled as a multivariate normal distribution.</p>
<p>The covariance matrix <span class="math inline">\(\mathbf{K}\)</span> is constructed using a Radial Basis Function (RBF) kernel, which defines the pairwise similarity between units <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> based on their distance:</p>
<p><span class="math display">\[
\begin{aligned}
K_{ij} &amp;= \eta^2 \exp\!\left( -\rho^2 D_{ij}^2 \right) + \delta_{ij}\epsilon \\
\end{aligned}
\]</span></p>
<p>In this kernel, <span class="math inline">\(\eta\)</span> is the marginal standard deviation of the Gaussian Process, and <span class="math inline">\(\rho\)</span> is the length-scale parameter controlling how correlation decays with increasing or decreasing distance.<br>
<span class="math inline">\(D_{ij}\)</span> represents the distance between units <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>, and <span class="math inline">\(\epsilon\)</span> is a small positive constant added to the diagonal through <span class="math inline">\(\delta_{ij}\)</span>, the Kronecker delta (jitter), to improve numerical stability.</p>
<p>The values of <span class="math inline">\(X_1\)</span>, <span class="math inline">\(\alpha\)</span>, <span class="math inline">\(\beta\)</span>, <span class="math inline">\(\eta\)</span>, <span class="math inline">\(\rho\)</span>, and <span class="math inline">\(\sigma\)</span> were estimated as parameters.<br>
The parameters <span class="math inline">\(\alpha\)</span> and <span class="math inline">\(\beta\)</span> were assigned normal distributions, while the hyperparameters <span class="math inline">\(\eta\)</span>, <span class="math inline">\(\rho\)</span>, and the residual standard deviation <span class="math inline">\(\sigma\)</span> were assigned exponential distributions.</p>
</section>
<section id="stan-code-4" class="level2">
<h2 class="anchored" data-anchor-id="stan-code-4">5.2 Stan code</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5"><pre class="sourceCode stan code-with-copy"><code class="sourceCode stan"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>  <span class="kw">data</span> { <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>&gt; N;</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[N] tightness;</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[N] X1;</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="dt">matrix</span>[N, N] D;</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> mu_alpha;</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> tau_alpha;</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>    <span class="dt">real</span> mu_beta;</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span> tau_beta;</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>    <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; s_eta;</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; s_rho;</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>  <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; lambda_sigma;</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>, <span class="kw">upper</span>=<span class="dv">1</span>&gt; prior_only;</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>  <span class="kw">transformed data</span> {</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>    <span class="dt">real</span> delta = <span class="fl">1e-8</span>;</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="kw">parameters</span> {</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>    <span class="dt">real</span> alpha;</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>    <span class="dt">real</span> beta;</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>    <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; eta;</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>    <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; rho;</span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>    <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; sigma;</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>    <span class="dt">vector</span>[N] gamma;</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a><span class="kw">model</span> {</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a>  <span class="dt">matrix</span>[N, N] K;</span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span>:N) {</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>    K[i, i] = square(eta) + delta;</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (j <span class="cf">in</span> (i+<span class="dv">1</span>):N) {</span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>      <span class="dt">real</span> d = D[i, j];</span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>      K[i, j] = square(eta) * exp( - square(rho) * square(d) );</span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>      K[j, i] = K[i, j]; }</span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a>      gamma ~ multi_normal(rep_vector(<span class="dv">0</span>, N), K);</span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a>      alpha ~ normal(mu_alpha, tau_alpha);</span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a>      beta ~ normal(mu_beta, tau_beta);</span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a>      eta ~ normal(<span class="dv">0</span>, s_eta);</span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a>      rho ~ normal(<span class="dv">0</span>, s_rho);</span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a>      sigma ~ exponential(lambda_sigma);</span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a>      <span class="dt">vector</span>[N] mu;</span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a>      <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span>:N) {</span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a>        mu[i] = alpha + beta * X1[i] + gamma[i]; }</span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (prior_only == <span class="dv">0</span>) {</span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true" tabindex="-1"></a>          tightness ~ normal(mu, sigma);</span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true" tabindex="-1"></a>          }</span>
<span id="cb5-52"><a href="#cb5-52" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-53"><a href="#cb5-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-54"><a href="#cb5-54" aria-hidden="true" tabindex="-1"></a><span class="kw">generated quantities</span> {</span>
<span id="cb5-55"><a href="#cb5-55" aria-hidden="true" tabindex="-1"></a>  <span class="dt">vector</span>[N] y_rep; <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span>:N) {</span>
<span id="cb5-56"><a href="#cb5-56" aria-hidden="true" tabindex="-1"></a>    y_rep[i] = normal_rng(alpha + gamma[i] + beta * X1[i], sigma);</span>
<span id="cb5-57"><a href="#cb5-57" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb5-58"><a href="#cb5-58" aria-hidden="true" tabindex="-1"></a>  }</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
</section>
<section id="additional-hierarchical-bayesian-linear-model-with-random-effects-covariance-structure" class="level1">
<h1>Additional: Hierarchical Bayesian Linear Model with Random-Effects Covariance Structure</h1>
<p>In completlly different analysis I have used similar but different model with brms package in R instead of the Stan code.<br>
Here we have contact (interpersonal contact with other individual with lesbian, gay, and bisexual person; LGB) and law enforcement (of same-sex marriage).</p>
<section id="mathematical-equation-5" class="level2">
<h2 class="anchored" data-anchor-id="mathematical-equation-5">Mathematical equation</h2>
<p><span class="math display">\[
\begin{aligned}
y_i &amp;\sim \mathrm{Normal}(\mu_i, \sigma) \\[4pt]
\mu_i &amp;= \alpha + \beta_1\,\text{Contact}_i + \beta_2\,\text{Law}_i
        + u^{\text{geo}}_{j[i]} + u^{\text{phy}}_{k[i]} \\[6pt]
\mathbf{u}^{\text{geo}} &amp;\sim \mathrm{MVN}(\mathbf{0}, \mathbf{\Sigma_G}) \\[4pt]
\mathbf{u}^{\text{phy}} &amp;\sim \mathrm{MVN}(\mathbf{0}, \mathbf{\Sigma_S}) \\[6pt]
\alpha &amp;\sim \mathrm{Normal}(0, \delta_\alpha) \\[4pt]
\beta_1 &amp;\sim \mathrm{Normal}(0, \delta_{\beta_1}) \\[4pt]
\beta_2 &amp;\sim \mathrm{Normal}(0, \delta_{\beta_2}) \\[6pt]
\delta_{\beta_1},\, \delta_{\beta_2} &amp;\sim \mathrm{Exponential}(\lambda_\beta) \\[4pt]
\sigma &amp;\sim \mathrm{Exponential}(\lambda_\sigma)
\end{aligned}
\]</span></p>
<p><span class="math inline">\(y_i\)</span> represents the outcome variable (e.g., justifiability of LGB) for individual <span class="math inline">\(i\)</span>,<br>
and <span class="math inline">\(Contact_i\)</span> and <span class="math inline">\(Law_i\)</span> are the individual- and country-level predictors, respectively.<br>
<span class="math inline">\(\alpha\)</span> is the overall intercept, while <span class="math inline">\(\beta_1\)</span> and <span class="math inline">\(\beta_2\)</span> represent the fixed-effect slopes for <span class="math inline">\(Contact_i\)</span> and <span class="math inline">\(Law_i\)</span>.</p>
<p>Importantly, <span class="math inline">\(u^{\text{geo}}_{j[i]}\)</span> and <span class="math inline">\(u^{\text{phy}}_{k[i]}\)</span> are structured random effects capturing country-level dependencies<br>
based on geographic and phylogenetic proximity, respectively.<br>
Each random-effect vector, <span class="math inline">\(\mathbf{u}^{\text{geo}}\)</span> and <span class="math inline">\(\mathbf{u}^{\text{phy}}\)</span>,<br>
is modeled as a multivariate normal distribution with zero mean and covariance matrices <span class="math inline">\(\boldsymbol{\Sigma_G}\)</span> and <span class="math inline">\(\boldsymbol{\Sigma_S}\)</span>.<br>
These covariance matrices represent the spatial correlation structure across countries, derived from geographic distance and cultural/phylogenetic similarity matrices.</p>
</section>
<section id="brms-code" class="level2">
<h2 class="anchored" data-anchor-id="brms-code">brms code</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">brm</span>(</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">formula =</span> outcome <span class="sc">~</span> contact <span class="sc">+</span> law <span class="sc">+</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    (<span class="dv">1</span> <span class="sc">|</span> <span class="fu">gr</span>(country_geo, <span class="at">cov =</span> G, <span class="at">id =</span> <span class="st">"geo"</span>)) <span class="sc">+</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    (<span class="dv">1</span> <span class="sc">|</span> <span class="fu">gr</span>(country_phy, <span class="at">cov =</span> S, <span class="at">id =</span> <span class="st">"phy"</span>)) ,</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> data,</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">data2 =</span> <span class="fu">list</span>(<span class="at">G =</span> G_pd, <span class="at">S =</span> S_pd),</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> <span class="fu">gaussian</span>(),</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">prior =</span> <span class="fu">c</span>(</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="at">class =</span> Intercept),</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">prior</span>(<span class="fu">normal</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="at">class =</span> b),</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">2</span>), <span class="at">class =</span> sd),</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">prior</span>(<span class="fu">exponential</span>(<span class="dv">1</span>), <span class="at">class =</span> sigma)),</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">cores =</span> <span class="dv">4</span>, <span class="at">iter =</span> <span class="dv">5000</span>, <span class="at">warmup =</span> <span class="dv">1000</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>

</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>